src/Language/DifferentialDatalog/Expr.hs

```diff
@@ -336,6 +338,30 @@ exprIsVarOrField' (EField _ e _)   = e
 exprIsVarOrField' (ETyped _ e _)   = e
 exprIsVarOrField' _                = False
 
+-- | Expression maps distinct assignments to input variables 'vs'
+-- to distinct outputs.
+exprIsInjective :: DatalogProgram -> S.Set String -> Expr -> Bool
+exprIsInjective d vs e =
+    S.fromList (exprVars e) == vs &&
+    exprFold (exprIsInjective' d) e
+
+-- No clever analysis here; just the obvious cases.
+exprIsInjective' :: DatalogProgram -> ExprNode Bool -> Bool
+exprIsInjective' _ EVar{}        = True
+exprIsInjective' d EApply{..}    =
+    and exprArgs && (maybe False (exprIsInjective d (S.fromList $ map name funcArgs)) $ funcDef)
```

Comment:
Is this check also applied to the function body in a function definition? If so this case is redundant, and may trigger non-termination if DDlog is to support recursive functions in the future.

Reply:
This is a good point. We would have to track visited functions to terminate the recursion.

Reply:
In fact, reasoning about recursive functions is more subtle. I will just add a comment saying that, once we allow recursion (and have the infrastructure to check that the function is recursive), `exprIsInjective` should return `False` for such functions.